receiver/vcenterreceiver/client.go

```diff
@@ -0,0 +1,209 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vcenterreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/performance"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/vim25"
+	vt "github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/govmomi/vsan"
+	"github.com/vmware/govmomi/vsan/types"
+)
+
+// vcenterClient is a client that
+type vcenterClient struct {
+	moClient   *govmomi.Client
+	vimDriver  *vim25.Client
+	vsanDriver *vsan.Client
+	finder     *find.Finder
+	pc         *property.Collector
+	cfg        *Config
+}
+
+func newVmwarevcenterClient(c *Config) *vcenterClient {
+	return &vcenterClient{
+		cfg: c,
+	}
+}
+
+func (vc *vcenterClient) Connect(ctx context.Context) error {
+	if vc.moClient == nil {
+		sdkURL, err := vc.cfg.SDKUrl()
+		if err != nil {
+			return err
+		}
+		client, err := govmomi.NewClient(ctx, sdkURL, vc.cfg.MetricsConfig.Insecure)
+		if err != nil {
+			return fmt.Errorf("unable to connect to vSphere SDK on listed endpoint: %w", err)
+		}
+		tlsCfg, err := vc.cfg.MetricsConfig.LoadTLSConfig()
+		if err != nil {
+			return err
+		}
+		client.DefaultTransport().TLSClientConfig = tlsCfg
+		user := url.UserPassword(vc.cfg.MetricsConfig.Username, vc.cfg.MetricsConfig.Password)
+		err = client.Login(ctx, user)
+		if err != nil {
+			return fmt.Errorf("unable to login to vcenter sdk: %w", err)
+		}
+		vc.moClient = client
+		vc.vimDriver = client.Client
+		vc.pc = property.DefaultCollector(vc.vimDriver)
+		vc.finder = find.NewFinder(vc.vimDriver)
+	}
+	return nil
+}
+
+func (vc *vcenterClient) Disconnect(ctx context.Context) error {
+	if vc.moClient != nil {
+		return vc.moClient.Logout(ctx)
+	}
+	return nil
+}
+
+func (vc *vcenterClient) ConnectVSAN(ctx context.Context) error {
+	vsanDriver, err := vsan.NewClient(ctx, vc.vimDriver)
+	if err != nil {
+		return err
+	}
+	vc.vsanDriver = vsanDriver
+	return nil
+}
```

Comment:
It's difficult to keep track of what's going on with the "client" when it has another client that may or may not be connected. Should these just be separate clients that you interact with as needed?

Reply:
I guess my main argument for keeping them together is that authentication is mostly handled via the `*vim25.Client` client being passed around. The current way makes the scraper not have to know about the `vim25.Client` being passed around to other clients, but I'm happy to spike something out if you feel like its preferred!

If I do end up making that refactor would you like to see a `vsan_client.go, performance_client.go, property_client.go`?